working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+
+- callback is invoked post reclamation and can't resurrect the referenced
+object;
+- callback is not allowed to invoke any Dart code or call most of Embedding API
+methods;
+
+### Design constraints and requirements
+
+Here are the main requirements to finalization API:
+
+- finalization callbacks should not interrupt sequential execution of the
+program in a way which is observable by pure Dart code;
+- there should be a way to minimize the possibility of _premature finalization_,
+that is finalization of the wrapper object while native resources owned by it
+are still in use;
+- when using finalization API for managing native resources developer should be
+able to:
+  - rely on them when writing fully synchronous code;
+  - rely on finalization callbacks to be invoked when isolate is shutting down;
+
+The first requirements is self-explanatory: we would like to avoid departures
+from run-to-completion model that Dart currently provides for synchronous code.
+
+Consider now the following code:
+
+```dart
+/// [Wrapper] holds native [resource] and registers its instances for
+/// finalization. When finalization callback is called it will destroy
+/// corresponding [resource].
+class Wrapper {
+  /// Native resource held by the wrapper. Lifetime
+  /// of the wrapper determines lifetime of the [_resource].
+  final ffi.Pointer<ffi.Void> _resource;
+
+  void method() {
+    // [this] might be reclaimed by GC after evaluating
+    // [this.resource] because it is never used again
+    // and not kept live
+    useResource(this._resource, causeGC());
+  }
+}
+```
+
+In this example if finalization can occur synchronously when `causeGC` is called
+`this._resource` might be released prematurely because no references to `this`
+remain alive after `this._resource` is evaluated and saved as an outgoing
+argument. This means that a garbage collection which occurs within `causeGC()`
+might reclaim a `Wrapper` object which was previous pointed by `this` and
+consequently invoke finalization callback associated with it, which in turn
+would release the native resource.
+
+## Proposal
+
+We split the proposed finalization API into two parts:
+
+- `FinalizationRegistry` which:
+  - does not provide strong guarantees around promptness of finalization;
+  - does not impose any restrictions on objects, you could associate a
+    finalization action with;
+  - does not impose any restrictions on finalization actions, but invokes them
+    asynchronously;
+  - is implementable on the Web.
+- `NativeFinalizationRegistry` which:
+  - provides stronger guarantees around promptness of finalization;
+  - guarantees that finalization actions are invoked when isolate is shutting
+  down;
+  - is limited to objects which implement `Finalizable` interface;
+  - imposes restrictions on finalization actions to allow calling these actions
+  outside of Dart universe.
+
+The following classes are added to `dart:core` library
```

Comment:
Can go the other way too:
```dart
abtract class ProposedFinalizationRegistry<F> {
  final void Function(F) _callback;  
  final Expando<List<Future<void>>> _unregisters = Expando<List<Future<void>>>();
  final _registry = FinalizationRegistry();  
    
  ProposedFinalizationRegistry(void Function(F) callback) : _callback = callback; 

  void register(Object object, F finalizationToken, [Object? unregisterToken]) {
    var future = _registry.register(object);
    if (unregisterToken != null) {
      (_unregisters[unregisterToken] ??= []]).add(future);
    }    
    future.then((_) { 
      _callback(finalizationToken);  
    });
  }
    
  void unregister(Object unregisterToken) {
    var futures = _unregisters[unregisterToken];
    if (futures != null) {
      _unregisters[unregisterToken] = null;
      for (var future in futures) _registry.unregister(future);
    }
  }
}
```
so the designs are equivalent.

Reply:
That proposed API with `Future` makes the most common case (execute the same code in the callback every time an object in the registry is GCed) less efficient.  Compiling to JS there would be more overhead of using the Future version as it requires the extra layer or expandos and completers you mention to go from the JS API to the Future based API. I'd be fine with adding the MyRegistry to package:collection or similar as some syntactic sugar for cases where users don't care as much about performance and wait to trigger a different callback for each object.
One other note is it would be quite easy to accidentally leak memory using the Future based api.
For example:
```dart
void someMethod(someObject, MyRegistry registry) async {
   await registry.register(someObject);
   print("Object $someObject finalized!");
}
```
Looks reasonable but can never work. You'd need to know to instead hold onto some finalization token to avoid keeping `someObject` alive. For example, the following case would be valid assuming async/await is implemented in a way that `someObject` is not kept alive until the end of the method

```dart
void someMethod(someObject, MyRegistry registry) async {
   var finalizationToken = someObject.id;
   await registry.register(someObject);
   print("Object $finalizationToken finalized!");
}
```

Reply:
I find the proposal with `Future` interesting, but I am not sure it is worth it.

Reply:
It's definitely higher overhead, but not necessarily much if we use internal access to `_Future`, or a specialized subclass of `_Future`, and skip the `Completer` and mapping from future to completer. Then it's more like *one object* per registration.

```dart
class _FinalizationFuture extends _Future<void> {
  final _FinalizationRegistry _registry;
  _FinalizationFuture(this._registry);
}
class _FinalizationRegistry {
  late _JsFinalizationRegistry _jsRegistry = _JsFinalizationRegistry(this._callback);
  Future<void> register(Object o) {
    var result = _FinalizationFuture(this);
    _jsRegistry.register(o, result, result);
    return result;
  }
  void unregister(Future<void> future, [Future<void>? completeWith]) {
    if (future is _FinalizationRegistry && identical(this, future._registry)) {
      _jsRegistry.unregister(future);
      if (completeWith != null && future._canComplete) {
        future._completeWithFuture(completeWith);
      }
    }
}
```


You don't get the ability to group multiple registrations under the same unregister token, so you'd have to keep track of those yourself. You keep the future alive by having it as the finalization token, unless you manage to GC the registry/future cycles entirely.

I still think it's a more Dart-like API than the callback based one.
The callback based API is *low-level* asynchrony, which makes it fit a low-level library like `dart:isolate`, not `dart:async` or `dart;core`.

Reply:
(So same question again: Are we creating a low-level API that is intended for specialized framework/package authors to build upon, or high-level API intended directly for application authors? If the latter, we should use the `Future` API here, if the former, we should probably put it somewhere more out of the way than `dart:core` or even `dart:async`)

Reply:
This is intended mainly for framework/package authors and FFI users. The documentation should make it clear that "FinalizationRegistry is probably not the solution you should be using unless you have a problem you cannot solve any other way."

Reply:
I'll at least prefer to make a go at a `Future` based API as part of implementation, so we can compare. If we can make it as efficient as the plain registry, then I think it interacts better with the rest of the language. If we can't, then at least we know that.

Reply:
I'm not sure the ergonomics of the `Future` based API are better than the callback based API because I expect most users will want to have a single callback for all entries in the registry rather than awaiting each one differently. If we wanted a more Dart centric API than perhaps using `Stream` rather than Future would be better with the `FinalizationRegistry` surfacing a stream of finalization tokens. I don't advocate that approach either as users who wanted that API could build it on top of the callback based API anyway.

If given an API that returns `Future`, the first thing I assume as a user is that  will play well with `await` but in this case I believe it plays badly with `await`. For example:
```dart

final registry = FinalizationRegistry();
Future<void> someMethod(Object obj) async {
  await registry.register(obj);
  print("Obj was disposed"); // I believe this will never hit unless rules for what objects are captured in an async method changed.
}

void someMethod(Object obj) {
  await registry.register(obj).then((_) {
    print("Obj was disposed"); // This should work as now obj can GC.
  });
}
```

Reply:
If those two examples differ in behavior, then I definitely can't predict how it works (both has an `await` which may keep the parameters alive until the function ends. Shouldn't matter which future it's waiting on, unless you specifically recognize tail-`await`s.

If the future approach encourages code which doesn't actually work, then it is probably not viable. A stream of finalization events would be easy to build, but also more specialized.

All in all, it leaves the question about where to put the low-level callback-based registry, because I don't think it really fits into any of the current platform-agnostic `dart:` libraries (it might fit in `dart:isolate`, but that isn't available on the web).

Reply:
Making this visible on the web is important so I am in favor of creating a new platform-agnostic library. I agree it shouldn't be in a core library as this is special purpose and rarely recommended functionality.
How about `dart:weakref`? The entire library could warn you that you shouldn't use it unless you are confident there is no other viable solution to your problem which is a nice benefit over putting it in an existing library.
As you mentioned `FinalizationRegistry` is basically an optimization that can be built on top of what you can do with `WeakRef` + polling.
This also matches how TypeScript has chunked the analogous classes.
https://fossies.org/linux/TypeScript/lib/lib.esnext.weakref.d.ts
contains both WeakRef and FinalizationRegistry

Reply:
Changed proposal to add those classes to `dart:weakref` instead of `dart:core`. Also added `typedef WeakMap = Expando;` for a good measure.